src/Domain/CoordinateMaps/Frustum.cpp

```diff
@@ -0,0 +1,339 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Domain/CoordinateMaps/Frustum.hpp"
+
+#include "DataStructures/MakeWithValue.hpp"
+#include "DataStructures/Tensor/EagerMath/DeterminantAndInverse.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+
+namespace CoordinateMaps {
+Frustum::Frustum(const std::array<std::array<double, 2>, 4> face_vertices,
+                 const double lower_bound, const double upper_bound,
+                 const Direction<3> direction_of_frustum) noexcept
+    : x1_(face_vertices[0][0]),
+      y1_(face_vertices[0][1]),
+      x2_(face_vertices[1][0]),
+      y2_(face_vertices[1][1]),
+      x3_(face_vertices[2][0]),
+      y3_(face_vertices[2][1]),
+      x4_(face_vertices[3][0]),
+      y4_(face_vertices[3][1]),
+      lower_bound_(std::move(lower_bound)),
+      upper_bound_(std::move(upper_bound)),
+      direction_of_frustum_(std::move(direction_of_frustum)) {}
+
+template <typename T>
+std::array<std::decay_t<tt::remove_reference_wrapper_t<T>>, 3> Frustum::
+operator()(const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = std::decay_t<tt::remove_reference_wrapper_t<T>>;
+
+  const ReturnType& xi = source_coords[0];
+  const ReturnType& eta = source_coords[1];
+  const ReturnType& zeta = source_coords[2];
+
+  ReturnType physical_x =
+      0.25 * ((x1_ * (1.0 - xi) + x2_ * (1.0 + xi)) * (1.0 - zeta) +
+              (x3_ * (1.0 - xi) + x4_ * (1.0 + xi)) * (1.0 + zeta));
+  ReturnType physical_y =
+      0.25 * ((y1_ * (1.0 - eta) + y2_ * (1.0 + eta)) * (1.0 - zeta) +
+              (y3_ * (1.0 - eta) + y4_ * (1.0 + eta)) * (1.0 + zeta));
+  ReturnType physical_z =
+      0.5 * (lower_bound_ * (1.0 - zeta) + upper_bound_ * (1.0 + zeta));
+
+  // Assigns location of frustum based on passed "direction_of_frustum"
+  // Frustums on z axis:
+  if (direction_of_frustum_.axis() == Direction<3>::Axis::Zeta) {
```

Comment:
See if you can factor all these conditionals into a function and then use it in Wedge too